### ISSUE TITLE
Feature request: add rule for `unittest` - always call `super().<setUp|tearDown|setUpClass|tearDownClass>()`

### ISSUE THREAD

**Issue by AdrianB-sovo** [astral-sh/ruff#14009]
In my tests, I'd like to make sure that all my derived classes of `unittest.TestCase` call the corresponding `super()` method for the methods `setUp()`, `tearDown()`, `setUpClass()` and `tearDownClass()`.

Could a linter rule be added to Ruff for that?

**Comment by AlexWaygood** [comment]
This seems reasonable to me. It would be quite a strict rule, but I think the chance of it producing invalid suggestions would be pretty low. `unittest.TestCase` already defines these methods, so it should be possible to always call `super().setUp()` etc. on any subclass of `unittest.TestCase`.

**Comment by mscheifer** [comment]
What's the potential danger of not calling `super()` ? The examples in the official Python docs don't call it: https://docs.python.org/3.13/library/unittest.html#organizing-test-code

**Comment by MichaReiser** [comment]
I don't think there's any real downside of not calling `setUp` etc, because unit test documents that they do nothing and I see it as extremely unlikely that this will ever change because a) it would be a breaking change and b) the `unittest` framework has other means to run some code before or after every test. 

The only argument I see for enforcing calling the `super` method is that it's generally considered best practice to call `super` in overridden methods even if they do nothing today because it safeguards against missing `super` call if the parent class *does something in the future*. 

Considering that it's highly unlikely that `unittest` will ever customize `setUp` and `tearDown`, I think this rule is too restrictive without providing real value. However, one issue that I observed more than once when working on a large Java code base is that tests forgot to call the `super` methods when inheriting from *custom* test classes and the custom test class started to override `setUp` (or `tearDown`) but not all subclasses were updated to call `super.setUp`. That's why I think the rule should instead lint for custom test classes and only then enforce that `setUp`, `tearDown` etc. calls the `super` method if overriden.

**Comment by AlexWaygood** [comment]
It's possible that we _could_ implement a version of this rule now that is only able to detect the indirect subclass of `TestCase` when the direct subclass is defined in the same file as the indirect subclass. That _does_ happen sometimes -- for example, https://github.com/python/cpython/blob/6ff8f82f92a8af363b2bdd8bbaba5845eef430fc/Lib/test/test__interpreters.py#L554-L557. It goes back to the question we've had to answer before about how much of an issue lots of false negatives are while we wait for multifile analysis... I honestly don't really have a sense for how common it is for indirect subclasses to be in the same file versus a different file as the direct subclass.

**Comment by njhearp** [comment]
I decided to close my pull request because it seems like this rule will need multi-file analysis to be useful and effective.

**Comment by AdrianB-sovo** [comment]
Yes, I also think this rule wouldn't be that useful without multi-file analysis.

Also, to confirm: the use case is when having multiple inherited levels of custom test classes (e.g. `unittest.TestCase` → `TestWithDB` → `BaseIntegrationTest` → `MySpecificTest`).

@MichaReiser 
> I think this rule is too restrictive without providing real value. [...] That's why I think the rule should instead lint for custom test classes and only then enforce that setUp, tearDown etc. calls the super method if overriden.

Yes, my description was ambiguous: we don't need to always have `<setUp|tearDown|setUpClass|tearDownClass>()` and the corresponding `super()` called in it.
I only want to enforce that the `super()` is called _**if**_ we implement them in our test classes.

**Comment by Michel-sss** [comment]
I'd like to take another pass at this with the narrower scope discussed here:

- **Do not** require `super()` on *direct* `unittest.TestCase` subclasses (no real upside, matches stdlib docs).
- **Do** require `super()` when overriding `setUp` / `tearDown` / `setUpClass` / `tearDownClass` on classes that inherit from a **custom** `TestCase` subclass that Ruff can resolve in the **same file** (Alex's same-file compromise until multifile analysis lands).
- Cross-module custom bases remain intentional false negatives for now.

Opening a PR shortly with preview rule `RUF077`.
